src/font.rs

```diff
@@ -114,6 +114,273 @@ impl<'a> Loca<'a> {
+        let &Encoding(data) = self;
+        get_u16(data, 6).unwrap()
+    }
+
+    fn get_seg_count(&'a self) -> u16 {
+        self.get_seg_count_x_2() / 2
+    }
+
+    fn get_search_range(&'a self) -> u16 {
+        let &Encoding(data) = self;
+        get_u16(data, 8).unwrap()
+    }
+
+    fn get_entry_selector(&'a self) -> u16 {
+        let &Encoding(data) = self;
+        get_u16(data, 10).unwrap()
+    }
+
+    fn get_range_shift(&'a self) -> u16 {
+        let &Encoding(data) = self;
+        get_u16(data, 12).unwrap()
+    }
+
+    fn get_u16_vec(&'a self, start_position: u16, count: u16) -> Vec<u16> {
+        let &Encoding(data) = self;
+        let mut result = vec![];
+        let mut vec_position = start_position;
+        let limit = vec_position + 2 * count;
+        while vec_position < limit {
+            result.push(get_u16(data, vec_position as usize).unwrap());
+            vec_position += 2;
+        }
+        result
+    }
+
+    fn get_i16_vec(&'a self, start_position: u16, count: u16) -> Vec<i16> {
+        let &Encoding(data) = self;
+        let mut result = vec![];
+        let mut vec_position = start_position;
+        let limit = vec_position + 2 * count;
+        while vec_position < limit {
+            result.push(get_i16(data, vec_position as usize).unwrap());
+            vec_position += 2;
+        }
+        result
+    }
+
+    fn get_end_counts_position() -> u16 {
+        14
+    }
+
+    fn get_end_counts(&'a self) -> Vec<u16> {
+        let seg_count = self.get_seg_count();
+        self.get_u16_vec(Self::get_end_counts_position(), seg_count)
+    }
+
+    fn get_start_counts_position(seg_count: u16) -> u16 {
+        Self::get_end_counts_position() + 2 + 2 * seg_count
+    }
+
+    fn get_start_counts(&'a self) -> Vec<u16> {
+        let seg_count = self.get_seg_count();
+        self.get_u16_vec(Self::get_start_counts_position(seg_count), seg_count)
+    }
+
+    fn get_id_deltas_position(seg_count: u16) -> u16 {
+        Self::get_start_counts_position(seg_count) + 2 * seg_count
+    }
+
+    fn get_id_deltas(&'a self) -> Vec<i16> {
+        let seg_count = self.get_seg_count();
+        self.get_i16_vec(Self::get_id_deltas_position(seg_count), seg_count)
+    }
+
+    fn get_id_range_offset_position(seg_count: u16) -> u16 {
+        Self::get_id_deltas_position(seg_count) + 2 * seg_count
+    }
+
+    fn get_id_range_offsets(&'a self) -> Vec<u16> {
+        let seg_count = self.get_seg_count();
+        self.get_u16_vec(Self::get_id_range_offset_position(seg_count), seg_count)
+    }
+
+    pub fn lookup_glyph_id(&'a self, code_point: u16) -> Option<u16> {
+        if self.get_format() != 4 {
+            return None;
+        }
+        let &Encoding(data) = self;
+        let end_counts_position = Self::get_end_counts_position();
+        let seg_count = self.get_seg_count();
+        let mut size = seg_count - 1;
+        let mut index = size / 2;
+        while size > 0 {
```

Comment:
Because this logic is a little hard to follow, it might be easier to separate out just the binary search, not mix it in with the (nontrivial) stuff that happens when found.